src/datasets/packaged_modules/base/autofolder.py

```diff
@@ -0,0 +1,398 @@
+import collections
+import itertools
+import os
+from dataclasses import dataclass
+from typing import ClassVar, List, Optional, Tuple
+
+import pyarrow.compute as pc
+import pyarrow.json as paj
+
+import datasets
+
+
+logger = datasets.utils.logging.get_logger(__name__)
+
+
+if datasets.config.PYARROW_VERSION.major >= 7:
+
+    def pa_table_to_pylist(table):
+        return table.to_pylist()
+
+else:
+
+    def pa_table_to_pylist(table):
+        keys = table.column_names
+        values = table.to_pydict().values()
+        return [{k: v for k, v in zip(keys, row_values)} for row_values in zip(*values)]
+
+
+def count_path_segments(path):
+    cnt = 0
+    while True:
+        parts = os.path.split(path)
+        if parts[0] == path:
+            break
+        elif parts[1] == path:
+            break
+        else:
+            path = parts[0]
+            cnt += 1
+    return cnt
+
+
+@dataclass
+class AutoFolderConfig(datasets.BuilderConfig):
+    """BuilderConfig for AutoFolder."""
+
+    _base_feature: ClassVar = (
+        None  # i.e. datasets.Image(), datasets.Audio(), we don't have a type for Feature ("FieldType")
+    )
+    base_feature_name: str = ""
+    label_column: str = "label"
+    features: Optional[datasets.Features] = None
+    drop_labels: bool = None
+    drop_metadata: bool = None
+
+    def __post_init__(self):
+        if not self.base_feature_name:
+            # automatically infer feature name from feature if it's not provided
+            if hasattr(self._base_feature, "_type"):
+                self.base_feature_name = self._base_feature._type.lower()
+            elif hasattr(self._base_feature, "__name__"):
+                self.base_feature_name = self._base_feature.__name__.lower()
+            else:
+                raise AttributeError(
+                    """
+                    `base_feature_name` is not provided and cannot be inferred form the base feature,
+                    check the config's `base_feature`.
+                    """
+                )
+
+
+class AutoFolder(datasets.GeneratorBasedBuilder):
+    BUILDER_CONFIG_CLASS = AutoFolderConfig
+    EXTENSIONS: List[str] = []
+    SKIP_CHECKSUM_COMPUTATION_BY_DEFAULT = True
+    METADATA_FILENAME: str = "metadata.jsonl"
```

Comment:
Feel free to separate the abstract attributes from the non-abstract ones (and add some comments). Moreover, in Python 3.6+, you can do `EXTENSIONS: List[str]` without specifying the default value to make it more obvious this is an abstract attribute.

Additionally, I think this class can be abstract (`class AutoFolder(datasets.GeneratorBasedBuilder, abc.ABC)`) to make sure calling `load_dataset`/`load_dataset_builder` fails on it.

P.S. I think `folder_builder`/`FolderBuilder` would be a better name for this module/class.

Reply:
`load_dataset` and `load_dataset_builder` won't work anyway because "autofolder" is not included in the `packaged_modules/__init__.py`. if it would, it wouldn't work with `config._base_feature` equals to `None`.

but I agree adding ABC makes the code more readable, but I'm not sure how to do it. just adding it as a base class doesn't make `inspect.isabstract()` return `True` as all abstract methods are implemented.

Reply:
> just adding it as a base class doesn't make inspect.isabstract() return True as all abstract methods are implemented.

I wasn't aware of that. 

Not marking this one as resolved as I'd like to hear @lhoestq's opinion on the naming (but I'm also ok with AutoFolder)

Reply:
@mariosasko you mean the name of the folder? i'll change it now, i just incorrectly understand your comment first, `FolderBuilder` looks better for me too

Reply:
[fe80766](https://github.com/huggingface/datasets/pull/4530/commits/fe80766ff7e58ff63d0a953c43536584a3b0ae78)

Reply:
"Auto" meant that something was happening automatically, which is not what the classe it doing here, it's just an abstract/base class. `FolderBuilder` is maybe not ideal given that it doesn't build folders... ^^'

I find `FolderBasedBuilder` nice, and is consistent with the naming of the other abstract/base builders that we have: `GeneratorBasedBuilder`, `ArrowBasedBuilder`, `BeamBasedBuilder`. What do you think ?

Reply:
@lhoestq I like `FolderBasedBuilder`. do you think we can preserve the name of directory and the script as `folder_builder/folder_builder.py` just to be shorter or is it not consistent and `folder_based_builder/folder_based_builder.py` is better?